# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/compiler/translator/tree_ops/RewriteDfdy.cpp
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/compiler/translator/tree_ops/RewriteDfdy.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/gfx/angle/checkout/src/compiler/translator/tree_ops/RewriteDfdy.cpp	Fri Nov 08 17:06:00 2019 +0000
@@ -0,0 +1,89 @@
+//
+// Copyright 2019 The ANGLE Project Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Implementation of dFdy viewport transformation.
+// See header for more info.
+
+#include "compiler/translator/tree_ops/RewriteDfdy.h"
+
+#include "common/angleutils.h"
+#include "compiler/translator/SymbolTable.h"
+#include "compiler/translator/tree_util/IntermNode_util.h"
+#include "compiler/translator/tree_util/IntermTraverse.h"
+
+namespace sh
+{
+
+namespace
+{
+
+class Traverser : public TIntermTraverser
+{
+  public:
+    static void Apply(TIntermNode *root,
+                      const TSymbolTable &symbolTable,
+                      TIntermBinary *viewportYScale);
+
+  private:
+    Traverser(TIntermBinary *viewportYScale, TSymbolTable *symbolTable);
+    bool visitUnary(Visit visit, TIntermUnary *node) override;
+
+    TIntermBinary *mViewportYScale = nullptr;
+};
+
+Traverser::Traverser(TIntermBinary *viewportYScale, TSymbolTable *symbolTable)
+    : TIntermTraverser(true, false, false, symbolTable), mViewportYScale(viewportYScale)
+{}
+
+// static
+void Traverser::Apply(TIntermNode *root,
+                      const TSymbolTable &symbolTable,
+                      TIntermBinary *viewportYScale)
+{
+    TSymbolTable *pSymbolTable = const_cast<TSymbolTable *>(&symbolTable);
+    Traverser traverser(viewportYScale, pSymbolTable);
+    root->traverse(&traverser);
+    traverser.updateTree();
+}
+
+bool Traverser::visitUnary(Visit visit, TIntermUnary *node)
+{
+    // Decide if the node represents a call to dFdy()
+    if (node->getOp() != EOpDFdy)
+    {
+        return true;
+    }
+
+    // Copy the dFdy node so we can replace it with the corrected value
+    TIntermUnary *newDfdy = node->deepCopy()->getAsUnaryNode();
+
+    size_t objectSize    = node->getType().getObjectSize();
+    TOperator multiplyOp = (objectSize == 1) ? EOpMul : EOpVectorTimesScalar;
+
+    // Correct dFdy()'s value:
+    // (dFdy() * ANGLEUniforms.viewportYScale)
+    TIntermBinary *correctedDfdy = new TIntermBinary(multiplyOp, newDfdy, mViewportYScale);
+
+    // Replace the old dFdy node with the new node that contains the corrected value
+    queueReplacement(correctedDfdy, OriginalNode::IS_DROPPED);
+
+    return true;
+}
+
+}  // anonymous namespace
+
+void RewriteDfdy(TIntermNode *root,
+                 const TSymbolTable &symbolTable,
+                 int shaderVersion,
+                 TIntermBinary *viewportYScale)
+{
+    // dFdy is only valid in GLSL 3.0 and later.
+    if (shaderVersion < 300)
+        return;
+
+    Traverser::Apply(root, symbolTable, viewportYScale);
+}
+
+}  // namespace sh
\ No newline at end of file